File: stack/infix_to_prefix_ll.py
# File Name: infix_to_prefix_ll.py

class Node:
    def __init__(self, data):
        self.data = data
        self.next = None

class Stack:
    def __init__(self):
        self.top = None

    def push(self, ch):
        node = Node(ch)
        node.next = self.top
        self.top = node

    def pop(self):
        ch = self.top.data
        self.top = self.top.next
        return ch

    def peek(self):
        return self.top.data

    def is_empty(self):
        return self.top is None

def precedence(ch):
    if ch in ['+', '-']:
        return 1
    if ch in ['*', '/']:
        return 2
    return 0

stack = Stack()
infix = input("Enter infix expression: ")

rev = ""
for ch in infix[::-1]:
    if ch == '(':
        rev += ')'
    elif ch == ')':
        rev += '('
    else:
        rev += ch

postfix = ""

for ch in rev:
    if ch.isalnum():
        postfix += ch
    elif ch == '(':
        stack.push(ch)
    elif ch == ')':
        while stack.peek() != '(':
            postfix += stack.pop()
        stack.pop()
    else:
        while not stack.is_empty() and precedence(stack.peek()) >= precedence(ch):
            postfix += stack.pop()
        stack.push(ch)

while not stack.is_empty():
    postfix += stack.pop()

prefix = postfix[::-1]
print("Prefix Expression:", prefix)

File: stack/InfixToPrefixLL.java
/* File Name: InfixToPrefixLL.java */

import java.util.Scanner;

class Node {
    char data;
    Node next;

    Node(char data) {
        this.data = data;
    }
}

class Stack {
    Node top = null;

    void push(char ch) {
        Node temp = new Node(ch);
        temp.next = top;
        top = temp;
    }

    char pop() {
        char ch = top.data;
        top = top.next;
        return ch;
    }

    char peek() {
        return top.data;
    }

    boolean isEmpty() {
        return top == null;
    }
}

public class InfixToPrefixLL {

    static int precedence(char ch) {
        if (ch == '+' || ch == '-') return 1;
        if (ch == '*' || ch == '/') return 2;
        return 0;
    }

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        Stack s = new Stack();

        System.out.print("Enter infix expression: ");
        String infix = sc.nextLine();

        String rev = "";
        for (int i = infix.length() - 1; i >= 0; i--) {
            char ch = infix.charAt(i);
            if (ch == '(') rev += ')';
            else if (ch == ')') rev += '(';
            else rev += ch;
        }

        String postfix = "";

        for (char ch : rev.toCharArray()) {
            if (Character.isLetterOrDigit(ch))
                postfix += ch;
            else if (ch == '(')
                s.push(ch);
            else if (ch == ')') {
                while (s.peek() != '(')
                    postfix += s.pop();
                s.pop();
            }
            else {
                while (!s.isEmpty() && precedence(s.peek()) >= precedence(ch))
                    postfix += s.pop();
                s.push(ch);
            }
        }

        while (!s.isEmpty())
            postfix += s.pop();

        String prefix = new StringBuilder(postfix).reverse().toString();
        System.out.println("Prefix Expression: " + prefix);
        sc.close();
    }
}
